Index: debianutils/ischroot.c
===================================================================
--- debianutils.orig/ischroot.c
+++ debianutils/ischroot.c
@@ -16,6 +16,7 @@
 #include <sys/types.h>
 #include <fcntl.h>
 #include <unistd.h>
+#include <errno.h>
 #ifdef HAVE_GETOPT_H
 #include <getopt.h>
 #endif /* HAVE_GETOPT_H */
@@ -135,13 +136,13 @@ static int ischroot()
   struct stat st1, st2;
   int ret;
 
-  ret = ischroot_mountinfo();
-  if (ret >= 0)
-    return ret;
-
   if (stat("/", &st1))
     return 2;
   if (stat("/proc/1/root", &st2)) {
+    /* If we get permission denied, try checking mountinfo instead,
+     * which can resolve mountinfo inconsistency issues in unshare environments */
+    if (errno == EACCES)
+      goto check_mountinfo;
     /* Does /proc/1/root exist at all? */
     if (lstat("/proc/1/root" , &st2))
       return 2;
@@ -154,6 +155,13 @@ static int ischroot()
     return 1;
   else
     return 0;
+
+check_mountinfo:
+  ret = ischroot_mountinfo();
+  if (ret >= 0)
+    return ret;
+
+  return 0;
 }
 
 #elif defined (__FreeBSD_kernel__) || defined (__FreeBSD__)
